File: libpandadna/DNAWall.py
import DNANode

from common import *

class DNAWall(DNANode.DNANode):
    COMPONENT_CODE = 10

    def __init__(self, name):
        DNANode.DNANode.__init__(self, name)

        self.code = ''
        self.height = 10
        self.color = (1, 1, 1, 1)

    def makeFromDGI(self, dgi):
        DNANode.DNANode.makeFromDGI(self, dgi)
        
        self.code = dgi_extract_string8(dgi)
        self.height = dgi.getInt16() / 100.0
        self.color = dgi_extract_color(dgi)
        
    def traverse(self, np, store):
        from DNAFlatBuilding import DNAFlatBuilding
        
        result = store.findNode(self.code)
        if result is None:
            self.raiseCodeNotFound()
            
        _np = result.copyTo(np)

        pos = Point3(self.pos)
        pos.setZ(DNAFlatBuilding.currentWallHeight)
        
        scale = Point3(self.scale)
        scale.setZ(self.height)

        _np.setPosHprScale(pos, self.hpr, scale)
        _np.setColor(self.color);

        self.traverseChildren(_np, store)

        DNAFlatBuilding.currentWallHeight += self.height
        

File: src/components/DNAWall.cxx
#include "DNAWall.h"
#include "DNAFlatBuilding.h"

TypeHandle DNAWall::_type_handle;

DNAWall::DNAWall(const std::string& name): DNANode(name), m_code(""),
                                           m_height(0), m_color(LVecBase4f(1))
{
}

DNAWall::~DNAWall()
{
}

void DNAWall::make_from_dgi(DatagramIterator& dgi, DNAStorage* store)
{
    DNANode::make_from_dgi(dgi, store);
    m_code = dgi.get_string();
    m_height = dgi.get_int16() / 100.0;
    m_color = DGI_EXTRACT_COLOR;
}

void DNAWall::traverse(NodePath& np, DNAStorage* store)
{
    NodePath result = store->find_node(m_code);
    if (result.is_empty())
    {
        raise_code_not_found();
        return;
    }
    
    NodePath _np = result.copy_to(np);
    
    LPoint3f pos = LPoint3f(m_pos);
    pos.set_z(DNAFlatBuilding::current_wall_height);

    LPoint3f scale = LPoint3f(m_scale);
    scale.set_z(m_height);
    
    _np.set_pos_hpr_scale(pos, m_hpr, scale);
    _np.set_color(m_color);
    
    traverse_children(_np, store);
    DNAFlatBuilding::current_wall_height += m_height;
}
